### PR TITLE
added support for case_weights

### PR DESCRIPTION
**Pull request by p-schaefer**
I noticed there wasn't much progress on integrating case_weights into lightgbm through bonsai so I've taken a crack at adding it. This is in reference to [65](https://github.com/tidymodels/bonsai/issues/65).

### CODE REVIEW

File: tests/testthat/test-lightgbm.R [hunk 1]
```diff
@@ -700,3 +700,109 @@ test_that("multi_predict() predicts classes if 'type' not given ", {
...
+                             , device = "cpu"
+                             , min_sum_hessian = 1e-4
+                             , num_leaves = 7L
+                             , nthread = 1L
+                             , min_data = 1L
+                             , valids = valids
+                             , seed = 1234L
+                             , deterministic = "true"
+                             , force_col_wise = "true"
+                             ) %>%
```
Comment (simonpcouch) [inline tests/testthat/test-lightgbm.R:765]: In addition to the style edits mentioned above, it'd be great if we could pare back the arguments set here as much as possible.

File: R/lightgbm.R [hunk 2]
```diff
@@ -35,7 +36,7 @@
 #' @return A fitted `lightgbm.Model` object.
 #' @keywords internal
 #' @export
-train_lightgbm <- function(x, y, max_depth = -1, num_iterations = 100, learning_rate = 0.1,
+train_lightgbm <- function(x, y,weights=NULL, max_depth = -1, num_iterations = 100, learning_rate = 0.1,
```
[the remark this answers is not in the document]
Comment (simonpcouch) [reply R/lightgbm.R:39]: https://style.tidyverse.org/syntax.html

**Review by simonpcouch — COMMENTED** [review]
A strong start! Some pointers to move this closer:

File: R/lightgbm.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (p-schaefer) [reply R/lightgbm.R]: I think I've corrected these

File: R/lightgbm.R [hunk 3]
```diff
@@ -224,18 +225,20 @@ process_data <- function(args, x, y, validation, missing_validation,
...
       categorical_feature = categorical_columns(x[trn_index, , drop = FALSE]),
-      params = list(feature_pre_filter = FALSE)
+      params = list(feature_pre_filter = FALSE),
+      weight = weights[trn_index]
     )
 
   if (!is.null(val_index)) {
     args$main$valids <-
       list(validation =
-          lightgbm::lgb.Dataset(
```
[the remark this answers is not in the document]
Comment (p-schaefer) [reply R/lightgbm.R:233]: I think I've corrected these? Let me know if thats not the case.

File: tests/testthat/test-lightgbm.R (hunk 1 above)
Comment (p-schaefer) [reply tests/testthat/test-lightgbm.R]: I pared the example down as suggested.

**Comment by p-schaefer** [comment]
Thanks for the review @simonpcouch. Note, I am also working on a method that allows some ability to customize validation data beyond just specifying a proportion of training data. I find this quite necessary for my work (since my data are not entirely independent - hence needing case weights as well), but I don't know if its robust enough to include in Bonsai proper.

File: R/lightgbm.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonpcouch) [reply R/lightgbm.R:233]: Almost there. If you look in this diff, there are several spaces before the function call and arguments that didn't used to be there.

File: R/lightgbm.R [hunk 4]
```diff
@@ -281,11 +284,11 @@ sort_args <- function(args) {
 # this function ensures that multiclass classification predictions are always
 # returned as a [num_observations, num_classes] matrix, regardless of lightgbm version
 reshape_lightgbm_multiclass_preds <- function(preds, num_rows) {
-    n_preds_per_case <- length(preds) / num_rows
```
Comment (simonpcouch) [inline R/lightgbm.R:284]: More space deletions here that are unrelated to the PR.

File: R/lightgbm.R [hunk 5]
```diff
@@ -324,9 +327,9 @@ predict_lightgbm_classification_class <- function(object, new_data, ...) {
 #' @rdname lightgbm_helpers
 predict_lightgbm_classification_raw <- function(object, new_data, ...) {
   if (using_newer_lightgbm_version()) {
-      p <- stats::predict(object$fit, prepare_df_lgbm(new_data), type = "raw", ...)
+    p <- stats::predict(object$fit, prepare_df_lgbm(new_data), type = "raw", ...)
   } else {
-      p <- stats::predict(object$fit, prepare_df_lgbm(new_data), rawscore = TRUE, ...)
```
Comment (simonpcouch) [inline R/lightgbm.R:329]: Here as well.

File: tests/testthat/_snaps/lightgbm.new.md
Comment (simonpcouch) [inline tests/testthat/_snaps/lightgbm.new.md:1]: This file, `lightgbm.new.md`, is an intermediate file from testthat. To merge its changes into the existing `test-lightgbm.R`, use `testthat::snapshot_accept()`.

File: tests/testthat/test-lightgbm.R (hunk 1 above)
Comment (simonpcouch) [reply tests/testthat/test-lightgbm.R]: Thanks!

**Comment by simonpcouch** [comment]
Thanks again for your work here. Given the lack of movement on this PR, I'm going to go ahead and close. :)

**Comment by simonpcouch** [comment]
That'd be wonderful, @joranE! Conceptually, the implementation here looks solid—following the current review comments would put this PR in a good spot.

**Comment by p-schaefer** [comment]
Sorry, this fell off my radar. I can finish addressing the comments next week, if that works for you.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by joranE** [comment]
@p-schaefer Sounds good, I'll let you tackle it! If you find yourself pulled away from it again for any reason, just let me know.

**Comment by p-schaefer** [comment]
I've fixed the spacing issues, and dealt with the intermediate file from testthat. Should I submit a new PR?

**Comment by simonpcouch** [comment]
You're welcome to push to the same branch! I'll reopen this one.

**Comment by p-schaefer** [comment]
Sorry, forgot to rerun the tests and checks. Looks good to me now, let me know if there is anything else.

**Comment by joranE** [comment]
Thanks for closing the loop on this @p-schaefer! Now I just need to figure out if there's an easy way to tweak this interface to allow for the use of the `linear_tree` parameter for piecewise linear trees, which "naturally" needs to be passed to `lgb.Dataset()` as well, instead of `lgb.train()`. 🙄

**Comment by joranE** [comment]
I added a new issue [here](https://github.com/tidymodels/bonsai/issues/77).

File: tests/testthat/test-lightgbm.R [hunk 6]
```diff
@@ -700,3 +700,74 @@ test_that("multi_predict() predicts classes if 'type' not given ", {
...
+  # https://github.com/microsoft/LightGBM/blob/638014d5c56fb92396bd55f344a47e3f651a3cd4/R-package/demo/weight_param.R
+
+  # Setup small weights
+  weights1 <- rep(1e-5, 6513L)
+  weights2 <- rep(1e-5, 1611L)
+
+  data(agaricus.train, package = "lightgbm")
+  dtrain <- lgb.Dataset(agaricus.train$data, label = agaricus.train$label, weight = weights1)
+  train <- data.frame(as.matrix(agaricus.train$data))
+  train$label<-agaricus.train$label
```
Comment (simonpcouch) [inline tests/testthat/test-lightgbm.R:725]: Could you do a pass through these tests enforcing [tidy style](https://style.tidyverse.org/syntax.html)?
```suggestion
  train$label <- agaricus.train$label
```

Namely, spaces before and after infix operators.

**Comment by simonpcouch** [comment]
Test failures are unrelated, feel free to ignore. :)

File: tests/testthat/test-lightgbm.R (hunk 6 above)
Comment (p-schaefer) [reply tests/testthat/test-lightgbm.R]: I hope I got them all this time.

**Comment by p-schaefer** [comment]
@simonpcouch Was there anything else you need from me?

**Comment by simonpcouch** [comment]
Nice work, @p-schaefer! :)
